### ISSUE TITLE
Thermostat set temperature not working

### ISSUE THREAD

**Issue by andrejkolesa** [nielsfaber/scheduler-card#288]
Hi,

i use Tuya TS0601_thermostats.

When i set in scheduler card Set temperature, nothing happens when timer expires.
I also tried to do customize and also does not work:

customize:
  climate.termostat_otroska:
    actions:
      - service: climate.set_temperature
        name: Set temperature1
        icon: fire
        variable:
          field: temperature
          name: Temperature
          min: 10
          max: 30
          step: 0.5
    exclude_actions:
      - heat
      - cool
      - manual

If i manually use climate.set_temperature and define temperature it works OK.

How can i debugg this problem ?

Thanks and br,
Andrej

**Comment by nielsfaber** [comment]
Hello Andrej,

Which version of the scheduler-card and scheduler-component do you use?
I released a new version of both today, so I recommend to update.
The customize should not be needed, I recommend to use the 'heat' / 'cool' actions.

Please create a schedule for 'heat' action to 20C which triggers in 5 minutes.
After that, check in the history of the schedule if it went to 'triggered' state at that time.
Then check in the history of your climate entity what happened at the same time.

**Comment by andrejkolesa** [comment]
Dear Niels,

I use latest version 2.0.3.
Under actions i do not have Heat. All i have available is:
-Set mode to Auto
-Set preset
-Set temprature
-Turn off

My attribures for climate are like this:
`hvac_modes: auto
min_temp: 5
max_temp: 35
target_temp_step: 0.5
preset_modes: none, away, schedule, manual, boost, complex, comfort, eco
current_temperature: 22.5
temperature: 22.5
preset_mode: manual
auto_lock: MANUAL
away_mode: OFF
away_preset_days: 1
away_preset_temperature: 15
battery_low: false
boost_time: 300
child_lock: LOCKED
comfort_temperature: 20
current_heating_setpoint: 22.5
eco_temperature: 15
force: normal
holidays: 
- hour: 6
  minute: 0
  temperature: 20
- hour: 136
  minute: 0
  temperature: 15
- hour: 11
  minute: 30
  temperature: 15
- hour: 12
  minute: 30
  temperature: 15
- hour: 17
  minute: 30
  temperature: 20
- hour: 22
  minute: 0
  temperature: 15

linkquality: 39
local_temperature: 22.5
local_temperature_calibration: 0
max_temperature: 35
min_temperature: 5
position: 0
preset: manual
system_mode: manual
update: 
state: idle

update_available: false
week: 5+2
window_detection: OFF
window_detection_params: 
minutes: 10
temperature: 5

workdays: 
- hour: 6
  minute: 0
  temperature: 20
- hour: 8
  minute: 0
  temperature: 15
- hour: 11
  minute: 30
  temperature: 15
- hour: 12
  minute: 30
  temperature: 15
- hour: 17
  minute: 30
  temperature: 20
- hour: 22
  minute: 0
  temperature: 15

friendly_name: termostat_dnevna1
supported_features: 17`

**Comment by andrejkolesa** [comment]
I use preset mode: manual

**Comment by nielsfaber** [comment]
Well, if your climate entity only supports hvac_mode=auto, which suggests:

> HVAC_MODE_AUTO | The device is set to a schedule, learned behavior, AI.

In other words, it does not allow you to control the temperature (the setpoint is controlled by the device internally).

The card checks for the property: 

> HVAC_MODE_HEAT | The device is set to heat to a target temperature.

If it is present, the card offers the capability to heat to a target value.

See here: https://developers.home-assistant.io/docs/core/entity/climate/#hvac-modes

It's weird that supported_features=17, since it suggests that is allows you to set a temperature, which is contradictive with the fact that you cannot control the HVAC_mode.

**Comment by nielsfaber** [comment]
You can still try if the 'set temperature' action works.
But I guess this will only work if the climate entity is in the correct preset.
Please share screenshots of the history of the schedule+climate entity after triggering, so we see what happens with both.

**Comment by andrejkolesa** [comment]
Set temperature sadly do not work.
I set it up for 17:40
Pictures of history are:
https://imgur.com/ZIcXfDe
https://imgur.com/Z478mCO

Schedule switch look like:

`weekdays: daily
times: 17:40:00
actions: 
- service: climate.set_temperature
  entity_id: climate.termostat_otroska
  service_data:
    temperature: 20

next_trigger: 2021-01-03T17:40:00+01:00
friendly_name: Schedule #a08f37
icon: mdi:calendar-clock`

And if i manually trigger this action, temperature is set on thermostat corectly.

**Comment by nielsfaber** [comment]
Could you try it using this panel:
![image](https://user-images.githubusercontent.com/16861605/103484144-d1a7fc80-4dec-11eb-9bb3-23a1906f8b3f.png)

It should have the same effect as the triggering of the scheduler.

**Comment by andrejkolesa** [comment]
This works correctly.

**Comment by andrejkolesa** [comment]
![climate_action](https://user-images.githubusercontent.com/74094100/103484985-42eaae00-4df3-11eb-9d81-dbea58eb6b05.PNG)
![climate_action_history](https://user-images.githubusercontent.com/74094100/103484987-467e3500-4df3-11eb-9bc7-b3e02359c057.PNG)

**Comment by nielsfaber** [comment]
Hello Andrej,
Do you still have this problem?
Or did you solve it in the mean time?

**Comment by andrejkolesa** [comment]
Dear Niels,

sadly problem is still there.
If you have any tips on how to debug i would apreciate it.

Thanks and br,
Andrej

**Comment by nielsfaber** [comment]
I have trouble pinpointing this issue.
Could anyone share their config of a schedule that gives this problem?
The config is stored in `scheduler.storage` in the `.storage`.
Thanks!

**Comment by andrejkolesa** [comment]
Here is my config:

https://drive.google.com/file/d/1QakF6YUz7F26LfwtBgKlZaxKM6V6-56V/view?usp=sharing

br,
Andrej

**Comment by andrejkolesa** [comment]
Good news Niels,

i tried again, and now it works OK.
In mean time Secheduler was updated as well as homeassistant (2021.1)

br,
Andrej

**Comment by danplant** [comment]
Hi. I'm also having a very similar issue.

I have schedules set up for all my z-wave trvs and UFH stat.

my scheduler.storage is here: https://drive.google.com/file/d/159ScRUWoNN7I5XAgETtwrYRgI9d3XA7O/view?usp=sharing

All the schedules are triggered but only the entities that have the hvac_mode attribute actually change. The other 3 (popp smart trv) don't react. They work fine when called manually or using an automation and also when called using an area. 

scheduler component is version v3.0.3
card is version v2.0.4
HA is version 2021.1.1

Not really sure what's happening.

Dan.

**Comment by linux-guru-com** [comment]
Same issue for me.
I figured out, that the component seems to be using the key 'service_data' instead of 'data'

![grafik](https://user-images.githubusercontent.com/58635259/105361748-4f944200-5bfa-11eb-8be8-50af5a8738ac.png)

When trying to create a "manual" automation with such syntax, it is ending up in an error complaining about exact this element (service_data).
IMHO the key should be here 'data', or am I wrong?

Cheers,
Tobi

**Comment by nielsfaber** [comment]
@danplant 
> All the schedules are triggered but only the entities that have the hvac_mode attribute actually change. The other 3 (popp smart trv) don't react. They work fine when called manually or using an automation and also when called using an area.

I'm not sure what's causing your problems.
Why is the hvac_mode "heat" not available for these entities?
Any device capable of heating a room should have it.
How do you normally operate these devices (switch between off and heating)?
Could it be that these entities have state `unavailable` or `unknown` (this would cause the scheduler to postpone the action)?
Your `scheduler.storage` file seems fine otherwise.

@linux-guru-com 
> I figured out, that the component seems to be using the key 'service_data' instead of 'data'

I'm not sure where you're going with this.
The screenshot you share is just the attributes of the schedule entity, it has no direct link with how data is sent to HA as service call.
In other words, the `service_data` is a chosen name, which could have been anything else.
Scheduler seems to work fine for 100+ users, so we should focus on the combination of scheduler + the specific device(s) you are using.

@danplant  @linux-guru-com 

Perhaps it could be helpful to enable debug logging for scheduler.
It should give more insight in what happens when the schedule is triggered.
You can do this by adding this to configuration.yaml (requires reboot of HA):
```
logger:
  default: warning
  logs:
    custom_components.scheduler: debug
```

**Comment by danplant** [comment]
> I'm not sure what's causing your problems.
> Why is the hvac_mode "heat" not available for these entities?
> Any device capable of heating a room should have it.
> How do you normally operate these devices (switch between off and heating)?
> Could it be that these entities have state `unavailable` or `unknown` (this would cause the scheduler to postpone the action)?
> Your `scheduler.storage` file seems fine otherwise.

Hi Niels. Thanbks for getting back to me. I forgot that I had to install a custom component to get these entities to show up properly in HA. There was a thread here: https://github.com/home-assistant/core/pull/29955

I had to reapply this fix when I last updated my python venv just before xmas. Anyway, these devices don't have HVAC heat mode. They are basic setpoint stats. I presume that somewhere herein lies the problem. I will turn on debugging as you suggest and see what I can find.

Thanks, again.

**Comment by linux-guru-com** [comment]
> 
> 
> @danplant
> 
> > All the schedules are triggered but only the entities that have the hvac_mode attribute actually change. The other 3 (popp smart trv) don't react. They work fine when called manually or using an automation and also when called using an area.
> 
> I'm not sure what's causing your problems.
> Why is the hvac_mode "heat" not available for these entities?
> Any device capable of heating a room should have it.
> How do you normally operate these devices (switch between off and heating)?
> Could it be that these entities have state `unavailable` or `unknown` (this would cause the scheduler to postpone the action)?
> Your `scheduler.storage` file seems fine otherwise.
 
... will refer to it in anwering the latest post of @danplant 

> In other words, the `service_data` is a chosen name, which could have been anything else.

Thanks for clarifying. It was not that clear to me, since I did not look into your code.
 
> Perhaps it could be helpful to enable debug logging for scheduler.

Did it already. Debug output is very basic. 😞 

> 2021-01-21 10:23:21 DEBUG (MainThread) [custom_components.scheduler] Scheduler coordinator is ready
2021-01-21 10:23:21 DEBUG (MainThread) [custom_components.scheduler.switch] Rescheduling timer for switch.schedule_d15e8f
2021-01-21 10:23:21 DEBUG (MainThread) [custom_components.scheduler.switch] The next timer is set for 2021-01-22T09:50:00+01:00
2021-01-21 10:23:21 DEBUG (MainThread) [custom_components.scheduler.switch] Rescheduling timer for switch.schedule_thermostate_wohnzimmer
2021-01-21 10:23:22 DEBUG (MainThread) [custom_components.scheduler.switch] We are starting in a timeslot. Proceed with actions.
2021-01-21 15:35:25 DEBUG (MainThread) [custom_components.scheduler.switch] Rescheduling timer for switch.schedule_d15e8f
2021-01-21 17:31:00 DEBUG (MainThread) [custom_components.scheduler.switch] Rescheduling timer for switch.schedule_thermostate_wohnzimmer

Cheers
Tobi

**Comment by linux-guru-com** [comment]
> 
> 
> > I'm not sure what's causing your problems.
> > Why is the hvac_mode "heat" not available for these entities?
> > Any device capable of heating a room should have it.
> > How do you normally operate these devices (switch between off and heating)?
> > Could it be that these entities have state `unavailable` or `unknown` (this would cause the scheduler to postpone the action)?
> > Your `scheduler.storage` file seems fine otherwise.
> 
> Hi Niels. Thanbks for getting back to me. I forgot that I had to install a custom component to get these entities to show up properly in HA. There was a thread here: [home-assistant/core#29955](https://github.com/home-assistant/core/pull/29955)
> 
> I had to reapply this fix when I last updated my python venv just before xmas. Anyway, these devices don't have HVAC heat mode. They are basic setpoint stats. I presume that somewhere herein lies the problem. I will turn on debugging as you suggest and see what I can find.
> 
> Thanks, again.

@nielsfaber 

I am using Devolo branded Danfoss thermostats called 'Devolo Home Control Radiator' and these devices do not use any hvac_mode as well.
![grafik](https://user-images.githubusercontent.com/58635259/105391656-9f830100-5c1a-11eb-9a4d-d4e5276830f5.png)

Seems, like th non existing hvac_mode and the therewith corresponding missing 'state' (here shown as 'unknown') is causing the issue. Maybe this gives you any direction where to look into your code?
I am happy to test any modifications.

Cheers
Tobi

**Comment by nielsfaber** [comment]
> Seems, like th non existing hvac_mode and the therewith corresponding missing 'state' (here shown as 'unknown') is causing the issue. Maybe this gives you any direction where to look into your code?
I am happy to test any modifications.

This is what’s giving you problems then.
Any action targeted at an entity having state `unknown` will be queued, since the scheduler integration assumes the entities are still being initialized.
This is intended functionality which is in place to be able to have scheduler execute tasks right after startup of HA (all entities start with state `unknown` until their integrations are ready).
I cannot help you further with this.

**Comment by danplant** [comment]
Okay.
So, I need to either 
a) get the entity to have some sort of state. Can I just force this in some way?
b) alter the scheduler code to ignore the entity state. Maybe as a selectable option.

Just out of curiosity, any pointers where I might start looking with regards to option b)?

Taa.

**Comment by linux-guru-com** [comment]
I'd second a potential workaround like @danplant stated with option b in https://github.com/nielsfaber/scheduler-card/issues/288#issuecomment-764956226 as I could imagine that there are a lot of thermostats around which are running potentially into the same issue.

What is the intention of your 'execute tasks right after startup' functionality respectively couln't it be handled by the 'homeassistant start' or another more generic trigger?

On Z-Wave connecitons there are possibilities to determine the 'ready' status as you could see on the screenshot, which maybe could be used for the 'unknown'/'right after startup' issue then. IMHO it would make only sense as a selectable switch for the exceptional behaviour.
E.g.: If switch 'no_hvac_mode' = on -> check ZWave(isready) and/or another indicator.

![grafik](https://user-images.githubusercontent.com/58635259/105496029-3607fe80-5cbd-11eb-9f70-e7373bb7eea4.png)

**Comment by danplant** [comment]
Okay. So I've commented out three sections of code in switch.py related to checking if the state is "unknown". This allows the action to get fired off rather than held indefinitely by the scheduler. Don't know if it'll break anything else but it seems to work on my system at the moment.
I'll look at making some sort of selectable option but it might take a while since I don't have a lot of time and, more importantly , I don't really know what I'm doing :)

Dan.

**Comment by linux-guru-com** [comment]
@danplant 
Do you mind sharing a diff or the line numbers which you commented out?
I'm absolutely unfamilar with Python... 😢

**Comment by danplant** [comment]
The modified file is here: https://drive.google.com/file/d/1npbqzYeeJfZfn8qp3uL9dBbBcsdb5hnx/view?usp=sharing

Don't know if this is the right way to share this. I suppose I should read up on using github properly.

**Comment by KTibow** [comment]
@danplant you can just drag+drop the file into the comment box, or put it the contents in a code box.

**Comment by danplant** [comment]
> @danplant you can just drag+drop the file into the comment box, or put it the contents in a code box.

Thanks. I'll remember that for the future.

**Comment by nielsfaber** [comment]
I am closing this issue since the discussion is going off-topic (and I'm cleaning up).
If anyone wants to make a case for having scheduler handle `unknown` state of entities, please open an issue in the scheduler-component (it has nothing to do with this card).

**Comment by danplant** [comment]
Okay Niels, I think that's sensible.

There wasn't that much interest in such a thing and I'm happy to run the scheduler with the slight modification I made for the time being.

Regards, Dan Plant. 

On 4 February 2021 17:25:30 GMT, Niels Faber <notifications@github.com> wrote:
>I am closing this issue since the discussion is going off-topic (and
>I'm cleaning up).
>If anyone wants to make a case for having scheduler handle `unknown`
>state of entities, please open an issue in the scheduler-component (it
>has nothing to do with this card).
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/nielsfaber/scheduler-card/issues/288#issuecomment-773476021

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.
